src/run-in-loop/index.ts

```diff
@@ -0,0 +1,89 @@
+import { go } from '@api3/promise-utils';
+
+import { type Logger } from '../logger';
+import { generateRandomBytes32, sleep } from '../utils';
+
+export interface RunInLoopOptions {
+  /** An API3 logger instance required to execute the callback with context. */
+  logger: Logger;
+  /** Part of every message logged by the logger. */
+  logLabel?: Lowercase<string>;
+  /**
+   * Minimum time to wait between executions. E.g. value 500 means that the next execution will be started only after
+   * 500ms from the start of the previous one (even if the previous one ended after 150ms). Default is 0.
+   */
+  frequencyMs?: number;
+  /**
+   * Minimum time to wait between executions. E.g. value 50 means that next execution will start 50ms after the previous
+   * one ended. Default is 0.
+   */
+  minWaitTime?: number;
+  /**
+   * Maximum time to wait between executions. E.g. value 100 means that the next execution will be started at most after
+   * 100ms from the end of the previous one. The maxWaitTime has higher precedence than minWaitTime of frequencyMs.
+   */
+  maxWaitTime?: number;
+  /**
+   * Maximum time to wait for the execution to finish. If the execution exceeds this time a warning is logged.
+   */
+  softTimeoutMs?: number;
+  /**
+   * Maximum time to wait for the execution to finish. If the execution exceeds this time an error is logged and the
+   * execution is force-stopped.
+   */
+  hardTimeoutMs?: number;
+  /**
+   * If false, the execution will not run. This is useful for temporarily disabling the execution (e.g. change
+   * environment variable and redeploy). Note that, there is no way to stop the loop execution once started. Default is
+   * true.
+   */
+  enabled?: boolean;
+}
+
+export const runInLoop = async (fn: () => Promise<void>, options: RunInLoopOptions) => {
+  const {
+    logger,
+    logLabel,
+    frequencyMs = 0,
+    minWaitTime = 0,
+    maxWaitTime,
+    softTimeoutMs,
+    hardTimeoutMs,
+    enabled = true,
+  } = options;
+
+  if (softTimeoutMs && hardTimeoutMs && hardTimeoutMs < softTimeoutMs) {
+    throw new Error('hardTimeoutMs must not be smaller than softTimeoutMs');
+  }
+
+  while (true) {
```

Comment:
Potential future improvement is to allow to stop this loop from running, (e.g return som object that allows you to call `terminate` or something like that), but this feature is not needed now.